### DISCUSSION TITLE
Получить играющий трек

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#66] [Идея] [answered]
Есть возможность получения текущего трека?
Сейчас история берется из Spotify, если я не ошибаюсь, раз в 15 минут. Недоигранный трек туда не попадает. Периодичность пополнения истории с Last fm надо полагать примерно такая же. Текущий трек туда попадет, но есть вероятность его подмены.
А в самом Spotify, на сколько я понимаю, есть API позволяющий увидеть текущий трек. Вижу его использование в других сервисах. Можно ли сделать такую функцию?
Есть идеи по ее применению, в динамических частообновляемых скриптах. Или например генерирование плейлиста по текущему треку, но без ввода ссылки на него.

**Comment by Chimildic** [comment, marked as answer]
В обновленной версии использовать [Player.getPlayback](https://chimildic.github.io/goofy/#/func?id=getplayback).

**Старый ответ**
Добавил функцию `RecentTracks.getPlayingTrack`. Нужно обновить [код](https://script.google.com/d/1DnC4H7yjqPV2unMZ_nmB-1bDSJT9wQUJ7Wq-ijF4Nc7Fl3qnbT0FkPSr/edit?usp=sharing) и [права доступа](https://chimildic.github.io/goofy/#/install?id=Обновить-права-доступа).

**Reply by Bedmit** [reply]
Что-то не работает. 
```
function RadioTrack1() {
  let Seed = RecentTracks.getPlayingTrack();
  let Radio = Source.craftTracks(Seed);

  Combiner.push(Seed, Radio);

  Playlist.saveAsNew({
  name: `${Seed[0].name} и т.п.`,
  tracks: Seed,
  description: Playlist.getDescription(Seed),
  sourceCover: Seed[0].album.images[0].url,
  public: false,
  });
}
```
выдает ошибку 
```
11:36:22	Ошибка	URL: https://api.spotify.com/v1/me/player/currently-playing 
Code: 401 
Content: {
  "error" : {
    "status" : 401,
    "message" : "Permissions missing"
  }
}
11:36:23	Ошибка	
TypeError: Cannot read property 'item' of undefined
getPlayingTrack	@ Library.gs:603
RadioTrack1	@ Test.gs:2
```

**Reply by Chimildic** [reply]
`Permissions missing` недостаточно прав

**Reply by Bedmit** [reply]
Еще раз обновил права доступа. Теперь так 
```
TypeError: tracks.map is not a function
mapToIds	@ Library.gs:447
createUrls	@ Library.gs:432
craftTracks	@ Library.gs:423
RadioTrack1	@ Test.gs:3
```

**Reply by Bedmit** [reply]
Заодно вопрос. Возможно использование двух вариантов в одном скрипте?
Сейчас у меня два варианта одного и того же скрипта. Один без ввода аргумента:
```
function RadioTrack1() {
  let Seed = RecentTracks.getPlayingTrack();
  let Radio = Source.craftTracks(Seed);
```
а второй с вводом ссылки на трек:
```
function RadioTrack2(values) {
  let Seed = SpotifyRequest.getFullObjByIds('tracks', [Helper.parseId(values.text.strTrackId)], 1);
  let Radio = Source.craftTracks(Seed);
```
Возможен вариант, когда скрипт один? Если в строку введена ссылка, то кнопка "Запуск" считает по ней. Если не введена, то по текущему треку.

В _launch_  оформил так:
```
{
                    name: 'RadioTrack1',
                    text: 'Радио по треку',
                }, 
                {
                    name: 'RadioTrack2',
                    args: [{ name: 'Ссылка на трек',key: 'strTrackId' }],
                }, 
```
В результате шрифт запуска функции без аргумента почему-то стал маленьким. Зато большим шрифтом показано название второй функции, хотя по логике его вообще не должно быть.

![2021-03-27](https://user-images.githubusercontent.com/76230194/112715958-994c2200-8ef4-11eb-93da-b20c17d50f04.png)

**Reply by Chimildic** [reply]
`craftTracks` принимает массив (один и более треков). `getPlayingTrack` возвращает объект (один трек). Нужно добавить квадратные скобки к `Seed`, чтобы создать массив из одного объекта. Аналогично сделать для `push`.
```js
let Radio = Source.craftTracks([Seed]);
```

> В результате шрифт запуска функции без аргумента почему-то стал маленьким

Не придумал как оформить команду без аргументов. Поэтому у них маленький шрифт.

> Возможен вариант, когда скрипт один?
```js
// launch.html
{
    name: 'createRadioByTrack',
    args: [{ name: 'Ссылка на трек', key: 'strTrackId' }],
}, 
```
```js
// Обработчик
function createRadioByTrack(values) {
  let seed;
  if (values.text.strTrackId.length > 0) {
      let id = Helper.parseId(values.text.strTrackId);
      seed = SpotifyRequest.getFullObjByIds('tracks', [id], 1)[0];
  } else {
      seed = RecentTracks.getPlayingTrack();
  }
  let radioTracks = Source.craftTracks([seed]);
}
```

**Reply by Bedmit** [reply]
Извиняюсь за долгое молчание. Уезжал.
В общем не выходит каменный цветок. Элементарно не хватает базовых знаний. Чтобы не расспрашивать по кусочкам, сразу прошу помощи в готовом скрипте.
```
function RadioTrack1() {
  let Seed = RecentTracks.getPlayingTrack();
  let Radio = Source.craftTracks([Seed]);

  let Past = Source.getSavedTracks();
  Filter.removeTracks(Radio, Past);
  let History = RecentTracks.get();
  Filter.removeTracks(Radio, History);
  Filter.rangeDateRel(History, 1);
  Filter.removeArtists(Radio, History); 
  Filter.dedupArtists(Radio);
  Filter.match(Radio, 'mix|club|radio|edit|live|karaoke', true);
  Order.sort(Radio, 'meta.popularity', 'desc');

  Combiner.push([Seed], Radio);

  Playlist.saveAsNew({
  name: `${Seed[0].name} &`,
  tracks: Seed,
  description: Playlist.getDescription(Seed),
  sourceCover: Seed[0].album.images[0].url,
  public: false,
  });
}


function RadioTrack2(values) {
  let Seed = SpotifyRequest.getFullObjByIds('tracks', [Helper.parseId(values.text.strTrackId)], 1);
  let Radio = Source.craftTracks(Seed);

  let Past = Source.getSavedTracks();
  Filter.removeTracks(Radio, Past);
  let History = RecentTracks.get();
  Filter.removeTracks(Radio, History);
  Filter.rangeDateRel(History, 1);
  Filter.removeArtists(Radio, History); 
  Filter.dedupArtists(Radio);
  Filter.match(Radio, 'mix|club|radio|edit|live|karaoke', true);
  Order.sort(Radio, 'meta.popularity', 'desc');
  
  Combiner.push(Seed, Radio);

  Playlist.saveAsNew({
  name: `${Seed[0].name} &`,
  tracks: Seed,
  description: Playlist.getDescription(Seed),
  sourceCover: Seed[0].album.images[0].url,
  public: false,
  });
}

```
Первая часть с текущим треком, так и не заработала. Подозреваю, где-то еще надо аргумент Seed брать в квадратные скобки. Вроде перебрал все варианты, но безрезультатно.

Задача та же - объединить все в один скрипт.

**Reply by Chimildic** [reply]
```js
// launch.html
{
  name: 'createRadioByTrack',
  args: [{ name: 'Ссылка на трек', key: 'strTrackId' }],
}, 
```
```js
// обработчик
function createRadioByTrack(values) {
  let track;
  if (values.text.strTrackId.length > 0) {
    let id = Helper.parseId(values.text.strTrackId);
    track = SpotifyRequest.getFullObjByIds('tracks', [id], 1)[0];
  } else {
    track = RecentTracks.getPlayingTrack();
  }
  
  let seed = [track];
  let radioTracks = Source.craftTracks(seed);
  let savedTracks = Source.getSavedTracks();
  let historyTracks = RecentTracks.get();

  Filter.dedupArtists(radioTracks);
  Filter.matchExcept(radioTracks, 'mix|club|radio|edit|live|karaoke');
  Filter.removeTracks(radioTracks, savedTracks);
  Filter.removeTracks(radioTracks, historyTracks);
  Filter.rangeDateRel(historyTracks, 1);
  Filter.removeArtists(radioTracks, historyTracks);

  Order.sort(radioTracks, 'meta.popularity', 'desc');
  let tracks = Combiner.push(seed, radioTracks);

  Playlist.saveAsNew({
    name: tracks[0].name,
    tracks: tracks,
    description: Playlist.getDescription(tracks),
    sourceCover: tracks[0].album.images[0].url,
    public: false,
  });
}
```

**Reply by Bedmit** [reply]
Да, теперь конечно все работает. Спасибо. Попробую сам разобраться, что у меня было не так.

**Reply by Bedmit** [reply]
В связи с недавним изменением функции `getPlayback`, перестал работать данный скрипт. При отсутствии ссылки на трек, запуск выдает ошибку. 
Замена `track = RecentTracks.getPlayingTrack();` на `track = RecentTracks.getPlayback();` проблему не решила. Возможно дело еще и в изменении файла `Helper`.
Все последние изменения применены. Прошу помощи. Собственных знаний исправить скрипт, не хватает.

**Reply by Chimildic** [reply]
```js
track = RecentTracks.getPlayback().item;
```

**Reply by Bedmit** [reply]
Спасибо. Заработало.
